src/slice/quicksort.rs

```diff
@@ -5,16 +5,20 @@
 //! `rayon_core::join`.
 
 use std::cmp;
+use std::marker::PhantomData;
 use std::mem::{self, MaybeUninit};
 use std::ptr;
 
 /// When dropped, copies from `src` into `dest`.
-struct CopyOnDrop<T> {
+struct CopyOnDrop<'src, 'dest, T> {
```

Comment:
Is it useful to have *two* lifetimes? They're both covariant, while invariant in `T` itself.

Reply:
I decided to go for full accuracy in the phantomdata, but we could technically have just one here

Reply:
I guess in a few places you're specifying explicit lifetimes parameters, otherwise NLL would probably tighten it to the life of `CopyOnDrop` itself.

Reply:
Would you like me to change it or leave it as is?

Reply:
Let's change to one for simplicity, thanks.

Reply:
How does this look? I could accept a third `&mut` ptr that ensures that dest is also factored in to the lifetime calculation but for now this seems to work.

Really wish there were a "raw pointer with a lifetime" type that had all the right methods but also the lifetime (but none of the validity constraints)